File: TRViS.UITests.Apple/Sources/Shared/OriginalTimetableV2PageObject.swift
// OriginalTimetableV2PageObject.swift
// Page object for the V2 (Card Stack) Original Timetable page.
// Mirrors TRViS.UITests/Pages/OriginalTimetableV2PageObject.cs.

import XCTest

class OriginalTimetableV2PageObject {

    private let app: XCUIApplication
    private let base: BaseUITestCase

    init(app: XCUIApplication, base: BaseUITestCase) {
        self.app = app
        self.base = base
    }

    // MARK: — Rendered state probe

    /// Waits up to `timeout` for V2 to reach a rendered state. Any of the
    /// tablet/compact header or empty-state anchors satisfies the wait.
    @discardableResult
    func waitForRendered(timeout: TimeInterval = 30) -> Bool {
        let anchors = [
            AutomationIds.OriginalTimetable.V2.header,
            AutomationIds.OriginalTimetable.V2.emptyState,
            AutomationIds.OriginalTimetable.V2.compactHeader,
            AutomationIds.OriginalTimetable.V2.compactEmptyState,
        ]
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            for id in anchors {
                if let _ = base.waitForElement(id: id, timeout: 0.5) {
                    return true
                }
            }
            Thread.sleep(forTimeInterval: 0.2)
        }
        return false
    }

    // MARK: — State predicates

    /// True when any V2 empty-state Label (tablet or compact) is in the tree.
    func isEmptyStateVisible(timeout: TimeInterval = 5) -> Bool {
        if base.waitForElement(
            id: AutomationIds.OriginalTimetable.V2.emptyState, timeout: timeout
        ) != nil { return true }
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V2.compactEmptyState, timeout: timeout
        ) != nil
    }

    /// True when the tablet sticky header is visible.
    func isHeaderVisible(timeout: TimeInterval = 5) -> Bool {
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V2.header, timeout: timeout
        ) != nil
    }

    /// True when ANY MarkerBadge (any row, V2-prefixed) is visible. Used by
    /// the negative pre-condition test.
    func isAnyMarkerBadgeVisible(timeout: TimeInterval = 5) -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            let predicate = NSPredicate(
                format: "identifier BEGINSWITH 'OriginalTimetable.V2.Row.' AND identifier ENDSWITH '.MarkerBadge'"
            )
            let query = app.descendants(matching: .any).matching(predicate)
            for i in 0..<min(query.count, 10) {
                let el = query.element(boundBy: i)
                if el.exists {
                    let frame = el.frame
                    if frame.size.width > 0 && frame.size.height > 0 {
                        return true
                    }
                }
            }
            Thread.sleep(forTimeInterval: 0.15)
        }
        return false
    }
}

File: TRViS.UITests/Pages/OriginalTimetableV2PageObject.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using TRViS.UITests.Infrastructure;

namespace TRViS.UITests.Pages;

/// <summary>
/// Page object for OriginalTimetableV2Page ("Card Stack"). Tablet layout
/// (width &gt;= 600pt) and compact layout are both implemented; the page is
/// "rendered" when EITHER the tablet sticky header, the compact header, or
/// either layout's empty-state Label is on screen.
/// </summary>
public class OriginalTimetableV2PageObject : PageObject
{
	public OriginalTimetableV2PageObject(AppiumDriver driver) : base(driver) { }

	public AppiumElement Header
		=> WaitForElement(AutomationIds.OriginalTimetable.V2.Header, TimeSpan.FromSeconds(30));
	public AppiumElement EmptyState
		=> WaitForElement(AutomationIds.OriginalTimetable.V2.EmptyState, TimeSpan.FromSeconds(30));
	public AppiumElement RowsList
		=> FindByAutomationId(AutomationIds.OriginalTimetable.V2.RowsList);

	/// <summary>
	/// Waits until V2 has finished its first render. The page is "rendered"
	/// when any of its tablet/compact header or empty-state anchors are
	/// reachable in the accessibility tree.
	/// </summary>
	public bool WaitForRendered(double timeoutSeconds = 30)
	{
		var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
		Driver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
		try
		{
			string[] anchors = new[]
			{
				AutomationIds.OriginalTimetable.V2.Header,
				AutomationIds.OriginalTimetable.V2.EmptyState,
				AutomationIds.OriginalTimetable.V2.CompactHeader,
				AutomationIds.OriginalTimetable.V2.CompactEmptyState,
			};
			while (DateTime.UtcNow < deadline)
			{
				foreach (var id in anchors)
				{
					try
					{
						if (Driver.FindElements(AutomationIdLocator(id)).Count > 0)
							return true;
					}
					catch { }
				}
				Thread.Sleep(200);
			}
			return false;
		}
		finally
		{
			Driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
		}
	}

	/// <summary>
	/// True when an empty-state Label (tablet OR compact) is currently visible.
	/// </summary>
	public bool IsEmptyStateVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V2.EmptyState, timeoutSeconds)
			|| PollDisplayed(AutomationIds.OriginalTimetable.V2.CompactEmptyState, timeoutSeconds);

	/// <summary>
	/// True when the tablet sticky header is currently visible — i.e. V2 is on
	/// the tablet layout AND has an ActiveTrain.
	/// </summary>
	public bool IsHeaderVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V2.Header, timeoutSeconds);

	/// <summary>True when the tablet layout grid is in the tree.</summary>
	public bool IsTabletLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V2.TabletGrid, timeoutSeconds);

	/// <summary>True when the compact layout grid is in the tree.</summary>
	public bool IsCompactLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V2.CompactGrid, timeoutSeconds);

	/// <summary>
	/// True when the marker badge for <paramref name="rowId"/> is visible.
	/// </summary>
	public bool IsMarkerBadgeVisible(string rowId, double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V2.MarkerBadgeFor(rowId), timeoutSeconds);
}
